tools/generator/templates/player/caml/interface.cc.jinja2

```diff
@@ -0,0 +1,247 @@
+{
+    CAMLparam1(in);
+    CAMLreturnT(int, Int_val(in));
+}
+
+template <>
+double caml_to_cxx<value, double>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT(double, Double_val(in));
+}
+
+template <>
+bool caml_to_cxx<value, bool>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT(bool, Int_val(in));
+}
+
+template <typename CxxType>
+std::vector<CxxType> caml_to_cxx_array(value in)
+{
+    CAMLparam1(in);
+    std::vector<CxxType> out;
+    mlsize_t size = Wosize_val(in);
+    for (size_t i = 0; i < size; ++i)
+        out.push_back(caml_to_cxx<value, CxxType>(Field(in, i)));
+    CAMLreturnT(std::vector<CxxType>, out);
+}
+{% for enum in game.enum %}
+
+{{ enum.enum_summary|cxx_comment }}
+template<>
+value cxx_to_caml<value, {{ enum.enum_name }}>({{ enum.enum_name }} in)
+{
+    CAMLparam0();
+    CAMLreturn(Val_int(in));
+}
+
+template<>
+{{ enum.enum_name }} caml_to_cxx<value, {{ enum.enum_name }}>(value in)
+{
+    CAMLparam1(in);
+    CAMLreturnT({{ enum.enum_name }}, ({{ enum.enum_name }})Int_val(in));
+}
+{% endfor %}
+{% for struct in game.struct %}
+
+{{ struct.str_summary|cxx_comment }}
+template<>
+value cxx_to_caml<value, {{ struct.str_name }}>({{ struct.str_name }} in)
+{
+    CAMLparam0();
+    CAMLlocal1(out);
+    out = caml_alloc({{ struct.str_field|length }}, 0);
+    {% for field_name, field_type, field_comment in struct.str_field %}
+    caml_initialize(&Field(out, {{ loop.index0 }}), {{ field_type|cxx_to_caml }}(in.{{ field_name }}));
+    {% endfor %}
+    CAMLreturn(out);
+}
+
+template<>
+{{ struct.str_name }} caml_to_cxx<value, {{ struct.str_name }}>(value in)
+{
+    CAMLparam1(in);
+    {{ struct.str_name }} out;
+    {% for field_name, field_type, field_comment in struct.str_field %}
+    out.{{ field_name }} = {{ field_type|caml_to_cxx }}(Field(in, {{ loop.index0 }}));
+    {% endfor %}
+    CAMLreturnT({{ struct.str_name }}, out);
+}
+{% endfor %}
+
+/*
+** Inititialize caml
+*/
+static inline void _init_caml()
+{
+    static bool is_initialized = false;
+    if (!is_initialized)
+    {
+        is_initialized = true;
+        char program_name[] = "./caml";
+        char* argv[2] = {program_name, NULL};
+        caml_startup(argv);
+    }
+}
+{% for func in game.function %}
+{%- set arg_list = func.fct_arg|caml_cxx_args %}
+
+{{ func.fct_summary|cxx_comment }}
+extern "C" value ml_{{ func.fct_name }}({# argument list #}
+                                        {%- set comma = joiner(', ') -%}
+                                        {%- for arg_name, _, _ in arg_list -%}
+                                        {{ comma() }}value {{ arg_name }}
+                                        {%- endfor -%}
+                                       )
+{
+    {# TODO: This only works for functions with 5 parameters or less.
+     # To handle more, we must call CAMLparam5 once, then CAMLxparamN, grouping
+     # the arguments 5 by 5 maximum #}
+    CAMLparam{{ arg_list|length }}({# argument list #}
```

Comment:
If the function doesn't have any parameter, we should be able to just call `CAMLparam0` and forget the `{%- set arg_list = func.fct_arg|caml_cxx_args %}`, I'll try that.